Shopify/ruby-lsp
# Do not append extra pipe if the user adds a closing pipe manually

### Motivation

If the user adds a closing pipe manually, we currently end up autocompleting in both cases and end up with 3 pipes.

This commits makes it so that if the user inserted a closing pipe to the end of the block argument, we remove the pipe we added automatically and avoid adding yet another one, resulting in two pipes.

refs: #838 

### Implementation

In this scenario, when the user adds a closing pipe manually to the end of the block argument, we end up with two pipes:

1. One in the current position where on type formatting was triggered
2. One in the current position - 1
3. And the opening pipe (and some block arguments)

If we're in this situation, we can remove the pipe we added through autocomplete so that we end up with only one pipe at the end.

### Automated Tests

Added a new example.

### Manual Tests

1. Start the LSP on this branch
2. Type `10.times do` in a file
3. Insert a pipe after the space
4. Verify the pipe is autocompleted
5. Insert a block argument `i`
6. Insert another pipe without moving the cursor
7. Verify that one of the extra pipes is removed and you end up with only one closing pipe

![ruby-lsp](https://github.com/Shopify/ruby-lsp/assets/748828/3a381947-9908-47db-ae33-4558d68709b4)
